### PR TITLE
For Ilias

### PR DESCRIPTION
**Pull request by churnfighter**

### CODE REVIEW

File: Tests/ChurnFighterTests/ThrottlerTests.swift [hunk 1]
```diff
@@ -0,0 +1,14 @@
+//
+//  ThrottlerTests.swift
+//  ChurnFighterTests
+//
+
+
+import XCTest
+@testable import ChurnFighter
+
+final class ThrottlerTests: XCTestCase {
```
Comment (ipavlidakis) [inline Tests/ChurnFighterTests/ThrottlerTests.swift:10]: maybe remove this file if it doesn't contain any tests?

File: Tests/ChurnFighterTests/Stubs/StubHttpClient.swift [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+@testable import ChurnFighter
+
+final class HttpError: Error {}
+
+final class StubHttpClient: HttpClientProtocol {
+    
+    private(set) var url: URL?
+    private(set) var jsonData: Data?
+    private(set) var isError: Bool = false
+    private(set) var data: Data = "".data(using: .utf8)!
```
Comment (ipavlidakis) [inline Tests/ChurnFighterTests/Stubs/StubHttpClient.swift:18]:
```suggestion
    private(set) var data: Data = Data()
```

File: Tests/ChurnFighterTests/ChurnFighterTests.swift [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+    private let userDefaults = UserDefaults(suiteName: "churnFighter")
+    private var mockBundleProvider: StubBundleProvider!
+    private var mockLocaleProvider: StubLocaleProvider!
+    private var mockDeviceProvider: StubDeviceProvider!
+    private var mockTimeZoneProvider: StubTimeZoneProvider!
+    
+    override func setUp() {
+        super.setUp()
+        
+        mockBundleProvider = StubBundleProvider(infoDictionary: ["CFBundleShortVersionString": "5.1.3"])
```
[the remark this answers is not in the document]
Comment (churnfighter) [reply Tests/ChurnFighterTests/ChurnFighterTests.swift:22]: I'm still confused after all these years.... tell me again!

File: Tests/ChurnFighterTests/ChurnFighterTests.swift [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+    func generateOfferEncodedString() -> String {
+        
+        let json = ["title": "offerTitle",
+                    "body": "offerBody",
+                    "productId": "com.product.id1",
+                    "offerId": "com.product.offer1",
+                    "offerType": "subscription",
+                    "productDescription": "productDescription"]
+        
+        let data = try! JSONSerialization.data(withJSONObject: json, options: [])
```
[the remark this answers is not in the document]
Comment (churnfighter) [reply Tests/ChurnFighterTests/ChurnFighterTests.swift:298]: old school :)

File: Sources/ChurnFighter/ApiClient.swift [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        let jsonObject = ["applicationUsername": usernameHash,
+                          "productIdentifier": productIdentifier,
+                          "offerIdentifier": offerIdentifier];
+        
+        if let json = try? JSONSerialization.data(withJSONObject: jsonObject, options: []),
+            let url = URL(string: "\(offerSignatureEndpoint)/\(apiKey)") {
+            
+            httpClient.post(url: url, jsonData: json) { (data, urlResponse, error) in
+                
+                if let result = try? JSONSerialization.jsonObject(with: data!, options: []) as? [String:AnyObject],
```
[the remark this answers is not in the document]
Comment (churnfighter) [reply Sources/ChurnFighter/ApiClient.swift:69]: but APIClient is internal, not public?

[the remark this answers is not in the document]
Comment (churnfighter) [reply Sources/ChurnFighter/ApiClient.swift:69]: do you mean the httpClient callback should return a struct?

File: Sources/ChurnFighter/ChurnFighter.swift [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+
+import Foundation
+import StoreKit
+import CommonCrypto
+
+
+@objc
+final public class ChurnFighter: NSObject {
+    
+    static private var privateShared: ChurnFighter? = nil
```
[the remark this answers is not in the document]
Comment (churnfighter) [reply Sources/ChurnFighter/ChurnFighter.swift:15]: mmm I don't fully understand - where do I store the shared instance? and when do I instanciate it? still in the configure method?

File: Sources/ChurnFighter/ChurnFighter.swift [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+    
+    private let bundleProvider: BundleProviding
+    private let deviceProvider: DeviceProviding
+    private let localeProvider: LocaleProviding
+    private let receiptHashKey = "receiptHash"
+    private var storeObserver = StoreObserver()
+    private let throttler = Throttler(seconds: 5)
+    private let timeZoneProvider: TimeZoneProviding
+    private let userDefaults = UserDefaults(suiteName: "churnFighter")
+    private var apiClient: ApiClient?
```
[the remark this answers is not in the document]
Comment (churnfighter) [reply Sources/ChurnFighter/ChurnFighter.swift:25]: that was my whole issue with doing

Churnfigther.configure(balh)

and then access Churnfighter.shared

What does the shared instance should be if we haven't called configure? In that case, I can't instanciate the APIclient without an API key  - what would you recommend?

File: Tests/ChurnFighterTests/ApiClientTests.swift [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+
+final class ApiClientTests: XCTestCase {
+    
+    private var apiClient: ApiClient!
+    private var mockHttpClient: StubHttpClient!
+    
+    override func setUp() {
+        super.setUp()
+        
+        mockHttpClient = StubHttpClient()
```
[the remark this answers is not in the document]
Comment (churnfighter) [reply Tests/ChurnFighterTests/ApiClientTests.swift:18]: 🤦‍♂️

File: Tests/ChurnFighterTests/ThrottlerTests.swift (hunk 1 above)
Comment (churnfighter) [reply Tests/ChurnFighterTests/ThrottlerTests.swift:10]: I was hoping you would have written tests for the Throttler class :)

**Pull request closed without merging**